src/errors.rs

```diff
@@ -0,0 +1,128 @@
+use base64::DecodeError;
+use std::{io, path::PathBuf, string::FromUtf8Error};
+use thiserror::Error;
+
+#[derive(Error, Clone, PartialEq)]
+pub enum TelemetryError {
+    //
+    // External errors
+    //
+    #[error("Base64 error: {0}")]
+    Base64(#[from] DecodeError),
+    #[error("IO error: {0}")]
+    IO(String),
+    #[error("Nix error: {0}")]
+    Nix(String),
+    #[error("Reqwest error: {0}")]
+    Reqwest(String),
+    #[error("Request clone failed")]
+    ReqwestCloneFailed,
+    #[error("UTF-8 error: {0}")]
+    Utf8(String),
+
+    //
+    // General errors
+    //
+    #[error("Invalid config: {0}")]
+    InvalidConfig(String),
+    #[error("Invalid client")]
+    InvalidClient,
+    #[error("Invalid 'RUST_LOG' filter: {0}")]
+    InvalidEnvFilter(String),
+    #[error("Home directory is invalid: {0}")]
+    InvalidHomeDir(String),
+    #[error("Log directory is invalid: {0}")]
+    InvalidLogDir(String),
+    #[error("Logfile is invalid: {0}/{1}")]
+    InvalidLogFile(String, PathBuf),
+    #[error("Invalid request")]
+    InvalidRequest,
+    #[error("Temporary directory is invalid: {0}")]
+    InvalidTmpDir(String),
+    #[error("Do not use 'TelemetryLayer::new()' directly, instead use the constructor macros")]
+    InvalidUsage,
+    #[error("Home directory is unreachable")]
+    UnreachableHomeDir,
+    #[error("Crate name is unreadable")]
+    UnreadableCrateName,
+
+    //
+    // FileWatcher errors
+    //,
+    #[error("Tracing event is invalid: {0}")]
+    InvalidTracingEvent(String),
+    #[error("Tracing regex is invalid: {0}")]
+    InvalidTracingRegex(#[from] regex::Error),
+    #[error("Tracing payload is invalid: {0}")]
+    InvalidTracingPayload(String),
+
+    // Mock error for testing
+    #[error("Mock error")]
+    Mock,
+}
+
+/// Convenience impl that dereferences the error
+/// so that we can use the error with `?`
+impl From<&TelemetryError> for TelemetryError {
```

Comment:
interesting pattern; this seems to be implicit behaviour.
Maybe we should just implement the `Copy` trait for this error type - that may achieve the same output?

Reply:
Yep that was what I initially did, but there are parameters to some of the variants that don't themselves implement `Copy` so it won't work.